### DISCUSSION TITLE
How is Preset Discovery initialization supposed to work, from a host's POV?

### DISCUSSION THREAD

**Discussion by Arakula** [free-audio/clap#295] [General]
The description in the header's header:
```
   The API works as follow to index presets and presets metadata:
   1. clap_plugin_entry.get_factory(CLAP_PRESET_DISCOVERY_FACTORY_ID)
   2. clap_preset_discovery_factory_t.create(...)
   3. ...
```
doesn't fully describe the functions exposed by the factory.

I _guess_ that clap_preset_discovery_factory.get_descriptor() should (could?) be used to determine which of the possibly multiple preset providers should be created. Is that about right?

I _guess_ that the provider ID passed to clap_preset_discovery_factory.create() should match the one used to create the Plugin in clap_plugin_factory.create_plugin(). Would make sense. It isn't documented, however. If I'm correct, then ... isn't get_descriptor() completely superfluous?

**Comment by abique** [comment]
You get the preset discovery factory from the clap entry.
It follows the same principle as the plugin.

**Reply by abique** [reply]
> I guess that the provider ID passed to clap_preset_discovery_factory.create() should match the one used to create the Plugin in clap_plugin_factory.create_plugin().

Not necessarily.

**Reply by Arakula** [reply]
> > I guess that the provider ID passed to clap_preset_discovery_factory.create() should match the one used to create the Plugin in clap_plugin_factory.create_plugin().
> 
> Not necessarily.

Then how should the host approach a situation where 2 or more PlugIns and/or 2 or more Preset Discovery Providers with different IDs could be created from the same .CLAP file? Simply create all Providers?

I mean, normally it will presumably be a simple 1:1 situation, but an advice how to deal with fringe cases would be a good idea.

**Reply by Arakula** [reply]
Guess I'm still not getting a clear view on the "big picture" here.
_When_ is the host supposed to do that?
As soon as it sees it (i.e., during the scan for new PlugIns)?
_Every_ time the host is opened?
Or when a CLAP PlugIn is instanciated?
